vissl/losses/barlow_twins_loss.py

```diff
@@ -0,0 +1,169 @@
+# Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved
+
+import pprint
+
+import torch
+from classy_vision.generic.distributed_util import (
+    all_reduce_mean,
+    get_cuda_device_index,
+)
+from classy_vision.losses import ClassyLoss, register_loss
+from torch import nn
+from vissl.utils.hydra_config import AttrDict
+
+
+@register_loss("barlow_twins_loss")
+class BarlowTwinsLoss(ClassyLoss):
+    """
+    This is the loss which was proposed in the Barlow Twins https://arxiv.org/abs/2103.03230v1
+    paper.
+    See the paper for the details on the loss.
+
+    Config params:
+        lambda_ (float):        weight on the off-diagonal terms. It controls the trade-off
+                                between the importance given to the invariance term versus the
+                                redundancy reduction term.
+        scale_loss (float):     In order to match the code that was used to develop Barlow
+                                Twins, we include an additional parameter, scale-loss, that
+                                multiplies the loss by a constant factor. We are working on a
+                                version that will not require this parameter.
+        embedding_dim (int):    dimensionality of the representation
+    """
+
+    def __init__(self, loss_config: AttrDict):
+        super(BarlowTwinsLoss, self).__init__()
+
+        self.loss_config = loss_config
+        self.bt_criterion = BarlowTwinsCriterion(
+            lambda_=self.loss_config.lambda_,
+            scale_loss=self.loss_config.scale_loss,
+            embedding_dim=self.loss_config.embedding_dim,
+        )
+
+    @classmethod
+    def from_config(cls, loss_config: AttrDict) -> "BarlowTwinsLoss":
+        """
+        Instantiates BarlowTwinsLoss from configuration.
+
+        Args:
+            loss_config:    configuration for the loss
+
+        Returns:
+            BarlowTwinsLoss instance.
+        """
+        return cls(loss_config)
+
+    def forward(self, output: torch.Tensor, target: torch.Tensor) -> torch.Tensor:
+        loss = self.bt_criterion(output)
+        return loss
+
+    def __repr__(self):
+        repr_dict = {
+            "name": self._get_name(),
+            "lambda_": self.loss_config.lambda_,
+            "scale_loss": self.loss_config.scale_loss,
+            "embedding_dim": self.loss_config.embedding_dim,
+        }
+        return pprint.pformat(repr_dict, indent=2)
+
+
+class BarlowTwinsCriterion(nn.Module):
+    """
+    The criterion corresponding to the Barlow Twins loss as defined in the paper
+    https://arxiv.org/abs/2103.03230v1.
+
+    Args:
+        lambda_ (float):        weight on the off-diagonal terms. It controls the trade-off
+                                between the importance given to the invariance term versus the
+                                redundancy reduction term.
+        scale_loss (float):     In order to match the code that was used to develop Barlow
```

Comment:
@jzbontar,

I saw in the official implementation that this parameter has been added. What purpose does it serve?

Reply:
The `--scale-loss` parameter multiplies the loss by a constant factor (1/32 by default). We include it in the code because we wanted to match our "research" code. I don't think scaling the loss is necessary, however, simply omitting the parameter and training with an unscaled loss results in a model that's a few percentage points worse than a model trained with a scaled loss. I'll try to remove the parameter from the code this week.

I think you can omit loss scaling from the VISSL implementation for now.

Reply:
is the loss scaling determined and dependent on the #gpus ?